File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/config/CacheConfig.java
package uk.gov.justice.digital.hmpps.keyworker.config;

import net.sf.ehcache.config.CacheConfiguration;
import net.sf.ehcache.store.MemoryStoreEvictionPolicy;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.CachingConfigurer;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.cache.ehcache.EhCacheCacheManager;
import org.springframework.cache.interceptor.CacheErrorHandler;
import org.springframework.cache.interceptor.CacheResolver;
import org.springframework.cache.interceptor.KeyGenerator;
import org.springframework.cache.interceptor.SimpleCacheErrorHandler;
import org.springframework.cache.interceptor.SimpleCacheResolver;
import org.springframework.cache.interceptor.SimpleKeyGenerator;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
@EnableCaching
public class CacheConfig implements CachingConfigurer {

    @Value("${cache.timeout.seconds.staff:86400}")
    private int staffInformationTimeoutSeconds;


    @Bean(destroyMethod = "shutdown")
    public net.sf.ehcache.CacheManager ehCacheManager() {
        final var config = new net.sf.ehcache.config.Configuration();

        config.addCache(config("getBasicKeyworkerDtoForStaffId", 10000, staffInformationTimeoutSeconds, MemoryStoreEvictionPolicy.LRU));
        config.addCache(config("getStaffDetailByUserId", 10000, staffInformationTimeoutSeconds, MemoryStoreEvictionPolicy.LRU));

        return net.sf.ehcache.CacheManager.newInstance(config);
    }

    public static CacheConfiguration config(final String name, final int maxElements, final int timeoutSeconds, final MemoryStoreEvictionPolicy policy) {
        return new CacheConfiguration().name(name)
                .memoryStoreEvictionPolicy(policy)
                .eternal(false)
                .overflowToOffHeap(false)
                .maxEntriesLocalHeap(maxElements)
                .timeToLiveSeconds(timeoutSeconds)
                .timeToIdleSeconds(timeoutSeconds);
    }

    @Bean
    @Override
    public CacheManager cacheManager() {
        return new EhCacheCacheManager(ehCacheManager());
    }

    @Bean
    @Override
    public KeyGenerator keyGenerator() {
        return new SimpleKeyGenerator();
    }

    @Bean
    @Override
    public CacheResolver cacheResolver() {
        return new SimpleCacheResolver(cacheManager());
    }

    @Bean
    @Override
    public CacheErrorHandler errorHandler() {
        return new SimpleCacheErrorHandler();
    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/config/CacheConfig.kt
package uk.gov.justice.digital.hmpps.keyworker.config

import net.sf.ehcache.CacheManager
import net.sf.ehcache.config.CacheConfiguration
import net.sf.ehcache.store.MemoryStoreEvictionPolicy
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.CachingConfigurer
import org.springframework.cache.annotation.EnableCaching
import org.springframework.cache.ehcache.EhCacheCacheManager
import org.springframework.cache.interceptor.*
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
@EnableCaching
class CacheConfig(@Value("\${cache.timeout.seconds.staff:86400}")
                  private val staffInformationTimeoutSeconds: Int = 0) : CachingConfigurer {

    @Bean(destroyMethod = "shutdown")
    fun ehCacheManager(): CacheManager {
        val config = net.sf.ehcache.config.Configuration()
        config.addCache(config("getBasicKeyworkerDtoForStaffId", 10000, staffInformationTimeoutSeconds, MemoryStoreEvictionPolicy.LRU))
        config.addCache(config("getStaffDetailByUserId", 10000, staffInformationTimeoutSeconds, MemoryStoreEvictionPolicy.LRU))
        return CacheManager.newInstance(config)
    }

    @Bean
    override fun cacheManager(): org.springframework.cache.CacheManager {
        return EhCacheCacheManager(ehCacheManager())
    }

    @Bean
    override fun keyGenerator(): KeyGenerator {
        return SimpleKeyGenerator()
    }

    @Bean
    override fun cacheResolver(): CacheResolver {
        return SimpleCacheResolver(cacheManager())
    }

    @Bean
    override fun errorHandler(): CacheErrorHandler {
        return SimpleCacheErrorHandler()
    }

    companion object {
        fun config(name: String?, maxElements: Int, timeoutSeconds: Int, policy: MemoryStoreEvictionPolicy?): CacheConfiguration {
            return CacheConfiguration().name(name)
                    .memoryStoreEvictionPolicy(policy)
                    .eternal(false)
                    .overflowToOffHeap(false)
                    .maxEntriesLocalHeap(maxElements)
                    .timeToLiveSeconds(timeoutSeconds.toLong())
                    .timeToIdleSeconds(timeoutSeconds.toLong())
        }
    }
}